### PR TITLE
[KDA] Add FlashInfer CAKE prefill and decode backends

### PR DESCRIPTION
**Pull request by yyihuang**
## Motivation

This PR integrates exported FlashInfer CAKE KDA kernels into SGLang and adds
an explicit `cake` backend for recurrent KDA prefill and decode:

```bash
--mamba-ssm-dtype bfloat16 --linear-attn-backend cake
```

The shared flag selects CAKE for both phases. Existing per-phase flags remain
authoritative when a mixed configuration is requested, and the verify backend
remains independently selected.

The serving-native packed decode API is being upstreamed in
[flashinfer-ai/flashinfer#4378](https://github.com/flashinfer-ai/flashinfer/pull/4378).
This PR depends on that PR landing and on SGLang consuming a FlashInfer
revision containing it. H=12 prefill support is already available through
[flashinfer-ai/flashinfer#4351](https://github.com/flashinfer-ai/flashinfer/pull/4351),
and the overall work is tracked in
[flashinfer-ai/flashinfer#4254](https://github.com/flashinfer-ai/flashinfer/issues/4254).

## Modifications

- Add `cake` to the linear-attention backend choices and dispatcher.
- Make `--linear-attn-backend cake` select CAKE for both prefill and decode,
  while preserving explicit per-phase overrides.
- Integrate CAKE prefill through `flashinfer.kda.recurrent_kda`.
- Integrate the Kimi-K3 T=1 serving hot path through the independent
  `flashinfer.packed_kda_decode` API.
- Forward SGLang's post-convolution packed QKV, raw gate/beta logits,
  caller-owned state pool, and original int32 state indices directly.
- Avoid Q/K/V splitting and copying, gate/beta preprocessing, and state
  gather/scatter on the covered packed-decode path.
- Preserve `-1` CUDA-graph padding rows and update active state slots in place.
- Log once when the serving-native packed CAKE route is actually exercised.
- Bypass Kimi-K3's separate fused decode path when `cake` is explicitly
  selected, ensuring that the requested backend is exercised.

## Packed decode contract and fallback

The CAKE packed path is selected only for the frozen FlashInfer contract:

- `T=1`, `H=HV=12`, `K=V=128`;
- BF16 packed QKV, raw gate/beta, output, and recurrent state;
- FP32 `A_log[12]` and `dt_bias[1536]`;
- packed QKV `[B, 4608]`;
- state pool `[N, 12, 128, 128]` with compact inner dimensions and a positive
  valid outer slot stride;
- contiguous CUDA int32 state indices `[B]`, with unique in-bounds active
  indices and `-1` for inactive graph rows;
- `scale=1/sqrt(128)`, `lower_bound=-5`, and `1 <= B <= 65535`;
- exact SM100 with CUDA >= 12.8 or SM103 with CUDA >= 12.9.

Unsupported shapes, dtypes, layouts, architectures, GQA, a missing packed
API, and ReplaySSM route through the existing Triton packed implementation.
Triton errors are propagated rather than hidden.

CAKE prefill similarly retains Triton for contracts requiring speculative or
state tracking, intermediate-state returns, T=1-only batches, unsupported
dimensions, or prefill CUDA-graph capture. Decode CUDA graphs are supported by
the packed CAKE path.

## Validation

Candidate revisions:

- SGLang PR head: `a6e3452c1cf42ada75fbc301ec99439651d1dfd2`;
- validated SGLang snapshot: `2c86a82dab900f4fb1589951e13a6cd8eb74bd68`;
- identical tracked tree for both SGLang revisions:
  `a50bae03fbfb97711d5189fa778155ab82f7e095`;
- FlashInfer performance commit:
  `89a5d3fcee8f8ddcff9da5547d095ffd73c1ef79`;
- FlashInfer #4378 head after non-kernel review cleanup:
  `826ed5f2fdc301440aa1ec799066f2fa3dfb63d5`.

GB300 / SM103 validation:

```text
FlashInfer JIT specification tests: 19 passed
FlashInfer packed KDA GPU tests:    26 passed
SGLang CPU adapter/selector tests:   9 passed
SGLang packed KDA GPU tests:         7 passed, 11 deselected
Pre-commit on changed files:         passed
```

The SGLang GPU tests cover B=1/8/64/128, the exact H=12 serving contract,
non-contiguous packed-QKV outer strides, non-identity state slots, `-1`
padding, padded state-pool outer strides, bitwise preservation of untouched
slots, forwarding of original state/index pointers, GQA fallback, and CUDA
Graph replay with changed inputs and indices through stable pointers.

FlashInfer #4378 additionally validates B=1/8/16/31/32/64/128/256/512 on B200
and GB300, current-stream execution, production strides, a 512-step FP64
diagnostic, and synccheck/memcheck with zero errors on both architectures.

### Exported public API vs exact SGLang packed Triton

Strict CUPTI, cold L2, 30 rounds per backend, same inputs, and alternating
AB/BA order:

| GPU | Direct B=1–512 geomean | CUDA Graph B=1–512 geomean |
|---|---:|---:|
| B200 / SM100 | `1.1058x` | `1.0863x` |
| GB300 / SM103 | `1.0708x` | `1.0526x` |

The complete per-batch tables, including retained low points, are in
FlashInfer #4378.

## Same-configuration Kimi-K3 E2E

Performance and accuracy use the same server launch configuration:

- Kimi-K3 TP8/DCP8 on 8-GPU GB300 allocations;
- BF16 recurrent state, HiCache, and full decode CUDA graphs;
- identical server and workload arguments apart from the selected
  linear-attention backend and unavoidable port values;
- random seed 42.

This is a same-configuration cross-run comparison, not a current-source paired
A/B in one allocation. The recorded Triton baseline used SGLang
`0a3efffc2f7fe8e0898473d86320d006973fdd71` and FlashInfer
`6b4ecd9d459caf389ce8c4a92bbb491214f2eac8` (Slurm `371396`). The CAKE run used
the validated SGLang snapshot
`2c86a82dab900f4fb1589951e13a6cd8eb74bd68` and FlashInfer
`826ed5f2fdc301440aa1ec799066f2fa3dfb63d5` (Slurm `378402`). Both used the
same GB300 SKU and the configuration above.

### Serving performance

Workload: 128 requests, 128 input tokens, 512 requested output tokens,
concurrency 64, greedy decoding, one warmup, cache flush before each measured
run, and three repetitions per backend.

| Metric | Triton runs | CAKE runs | Triton median | CAKE median | Speedup |
|---|---|---|---:|---:|---:|
| Decode output tok/s | `2008.276 / 1979.075 / 1959.839` | `1992.671 / 1985.439 / 1967.150` | `1979.075` | `1985.439` | `1.0032x` |
| Total tok/s | `2510.345 / 2473.844 / 2449.798` | `2490.838 / 2481.799 / 2458.938` | `2473.844` | `2481.799` | `1.0032x` |
| Median E2E latency (ms) | `16301.127 / 16542.684 / 16705.231` | `16432.781 / 16490.611 / 16645.532` | `16542.684` | `16490.611` | `1.0032x` |
| Median TTFT (ms) | `650.728 / 805.163 / 741.787` | `666.114 / 748.696 / 808.828` | `741.787` | `748.696` | `0.9908x` |
| Median TPOT (ms) | `30.656 / 30.806 / 30.934` | `30.794 / 30.814 / 30.987` | `30.806` | `30.814` | `0.9998x` |
| Median ITL (ms) | `30.459 / 30.661 / 30.799` | `30.402 / 30.642 / 30.821` | `30.661` | `30.642` | `1.0006x` |

All repetitions completed with zero request errors. The CAKE route log was
checked only after real generation requests had exercised the server.

All six runs completed 128/128 requests with the expected 16,384 input and
65,536 output tokens. The output-throughput median is nominally `1.0032x`, but
the decode TPOT median is effectively flat (`0.9998x`). Given the cross-run
method and this noise-scale delta, these data do not establish a material E2E
speedup.

### GSM8K accuracy

Protocol: 200 examples, five-shot completion API, greedy sampling, maximum 512
tokens, 128 client threads, and three repetitions against the same server
process used for the performance measurements.

| Triton runs | CAKE runs | Triton median | CAKE median | Delta | Gate |
|---|---|---:|---:|---:|---|
| `0.970 / 0.980 / 0.985` | `0.975 / 0.985 / 0.975` | `0.980` | `0.975` | `-0.005` | PASS |

The quality gate requires a CAKE median of at least `0.97` and a median
regression of no more than `0.01` relative to Triton.

## Checklist

- [x] Add the serving-native packed CAKE decode route.
- [x] Preserve explicit Triton fallbacks for unsupported contracts.
- [x] Add targeted CPU and GPU correctness tests.
- [x] Run pre-commit on all changed files.
- [ ] Merge FlashInfer #4378.
- [ ] Update SGLang's FlashInfer dependency to a revision containing #4378.
- [x] Report same-configuration Kimi-K3 performance and GSM8K results.

---
### CI States

Latest PR Test (Base): :x: [Run #31091488033](https://github.com/sgl-project/sglang/actions/runs/31091488033)
Latest PR Test (Extra): :no_entry_sign: [Run #31224311301](https://github.com/sgl-project/sglang/actions/runs/31224311301)

### CODE REVIEW

**Comment by yyihuang** [comment]
Closing in favor of #34299, which supersedes this PR and carries the current Phase-A implementation and validation evidence. Keeping this PR for historical review context.
